scapy/contrib/mqtt.py

```diff
@@ -0,0 +1,251 @@
+## This file is part of Scapy
+## See http://www.secdev.org/projects/scapy for more informations
+## Copyright (C) Santiago Hernandez Ramos <shramos@protonmail.com>
+## This program is published under GPLv2 license
+
+
+from scapy.packet import *
+from scapy.fields import *
+from scapy.layers.inet import TCP
+
+
+# CUSTOM FIELDS
+
+class VariableFieldLenField(FieldLenField):
+    def addfield(self, pkt, s, val):
+        val = self.i2m(pkt, val)
+        data = []
+        while val:
+            if val > 127:
+                data.append(val & 127)
+                val /= 127
+            else:
+                data.append(val)
+                lastoffset = len(data) - 1
+                data = "".join(chr(val | (0 if i == lastoffset else 128))
+                               for i, val in enumerate(data))
+                return s + data
+            if len(data) > 3:
+                raise Scapy_Exception("%s: malformed length field" %
+                                      self.__class__.__name__)
+    def getfield(self, pkt, s):
+        value = 0
+        for offset, curbyte in enumerate(s):
+            curbyte = ord(curbyte)
+            value += (curbyte & 127) * (128 ** offset)
+            if curbyte & 128 == 0:
+                return s[offset + 1:], value
+            if offset > 2:
+                raise Scapy_Exception("%s: malformed length field" %
+                                      self.__class__.__name__)
+
+
+
+# LAYERS
+            
+control_packet_types = {1:'CONNECT (1)',
+                        2:'CONNACK (2)',
+                        3:'PUBLISH (3)',
+                        4:'PUBACK (4)',
+                        5:'PUBREC (5)',
+                        6:'PUBREL (6)',
+                        7:'PUBCOMP (7)',
+                        8:'SUBSCRIBE (8)',
+                        9:'SUBACK (9)',
+                        10:'UNSUBSCRIBE (10)',
+                        11:'UNSUBACK (11)',
+                        12:'PINGREQ (12)',
+                        13:'PINGRESP (13)',
+                        14:'DISCONNECT (14)'}
+
+
+QOS_level = {0:'0 (At most once delivery)',
+             1:'1 (At least once delivery)',
+             2:'2 (Exactly once delivery)'}
+
+
+class MQTT(Packet):
+    name = "MQTT fixed header"
+    fields_desc = [
+        BitEnumField("type", 1, 4, control_packet_types),
+        BitEnumField("DUP", 0, 1, {0:'Disabled (0)',
+                                   1:'Enabled (1)'}),
+        BitEnumField("QOS", 0, 2, QOS_level),
+        BitEnumField("RETAIN", 0, 1, {0:'Disabled (0)',
+                                      1:'Enabled (1)'}),
+        VariableFieldLenField("len", None, length_of="len",
+                              adjust=lambda pkt, x: len(pkt.payload),),
```

Comment:
Same here about stackoverflow. Also, since this is a hack, a comment explaining it would be more than welcome.